### PR TITLE
Feat(challenge): Only send the challenges for the requested block

### LINKED ISSUE #13953: "Something went wrong error" primarily affecting campers around South Africa
I am trying to log in to my account at beta.freecodecamp.com and start the HTML challenges, but while i am always receiving this message: **SOMETHING WENT WRONG, PLEASE TRY AGAIN LATER**
i have tried many times, but it has not changed. 

on this page, while clicking to "Finish Challenge", i am still getting that same "SOMETHING WENT WRONG, PLEASE TRY AGAIN LATER"

how can i fix it?
beta issue

### LINKED ISSUE #16468: Facing Issue related to completing challenges. Getting something went wrong error.
#### Issue Description

Not able to go to next challenge or visit challenges in map. 

Map is totally unresponsive. Cannot visit any links.   Cannot complete and go to next challengesas well by pressing Ctrl-Enter or Clicking Modal. 

I get **Something went wrong** error everytime. Attached screenshots for better understanding. 

#### Browser Information

* Browser Name, Version: Firefox - 58.0b7 (64-bit), Chrome - 63.0.3239.132 (Official Build) (64-bit)
* Operating System: Ubuntu 17.10, 64bit 
* Mobile, Desktop, or Tablet: Desktop

#### Screenshot

fcc-beta-0
fcc-beta-1
fcc-beta-2
fcc-beta-3
fcc-chrome-scrshot

### LINKED ISSUE #16766: Beta site not working.
Beta site does not seems to work fine since a couple of days. I cannot go to the next challenge, or use map. Whenever I try to click map, i get this error - 
screen shot 2018-02-22 at 12 28 08 pm
And when I try to click go to next challenge, this error occurs - 
screen shot 2018-02-22 at 12 27 32 pm
Can someone look into this issue?
Operating system - MacOS / Chrome

### PR DESCRIPTION
**Pull request by Bouncey**
#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of freeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [ ] You have only one commit (if not, [squash](http://forum.freecodecamp.org/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change

- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [x] Tested changes locally.
- [x] Addressed currently open issue (replace XXXXX with an issue no in next line)

Closes #13953
Closes #16468
Closes #16766

#### Description

This PR adds the ability for the server to only send the challenges required for the users current block.

### CODE REVIEW

**Comment by QuincyLarson** [comment]
@Bouncey Wow - this is in some cases a 10x reduction in the amount of data that needs to be fetched. This should significantly speed up the app - especially for first-time users who don't have any thing cached.

@BerkeleyTrue I looked at this and didn't see any regressions - could you take a quick look at the code itself and merge this if it looks good to you?

File: common/app/redux/fetch-challenges-epic.js [hunk 1]
```diff
@@ -49,39 +62,84 @@ export function fetchChallengeEpic(actions, { getState }, { services }) {
...
+          { isDev }
+        );
+      }
+      const isNewBlockRequired = (
+        (isNewBlock || isNewSuperBlock) &&
+        !nextChallenge.description
+      );
+      return isNewBlockRequired ?
+        fetchNewBlock(nextChallenge.block) :
+        { type: 'NULL' };
```
Comment (BerkeleyTrue) [inline common/app/redux/fetch-challenges-epic.js]: Instead send just null and use a filter downstream. Don't dispatch and action when you don't need to.

File: common/app/redux/index.js [hunk 2]
```diff
@@ -66,7 +68,7 @@ const throwIfUndefined = () => {
 //   label?: String,
 //   value?: Number
 // }) => () => Object
-export const createEventMetaCreator = ({
+export function createEventMetaCreator({
```
Comment (BerkeleyTrue) [inline common/app/redux/index.js]: Why is this change made?

File: server/services/challenge.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      ) {
+      const getChallengeBlock$ = challengeMap.map(getMapForLang(lang))
+        .flatMap(({
+          result: { superBlocks },
+          entities: {
+            block: fullBlockMap,
+            challenge: challengeMap
+          }
+        }) => {
+          log(`sourcing challenges for the ${blockName} block`);
```
Comment (BerkeleyTrue) [inline server/services/challenge.js]: The indenting here for the rest of the function body is off

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]
Looks solid. A couple of nit picks.

File: common/app/Map/Map.jsx [hunk 4]
```diff
@@ -4,23 +4,32 @@ import { connect } from 'react-redux';
...
 
 export class ShowMap extends PureComponent {
-  renderSuperBlocks(superBlocks) {
+
+  renderSuperBlocks() {
+    const { superBlocks } = this.props;
     if (!Array.isArray(superBlocks) || !superBlocks.length) {
-      return <div>No Super Blocks</div>;
+      return (
+        <div style={{ hieght: '300px' }}>
```
Comment (ojongerius) [inline common/app/Map/Map.jsx]: Typo: "hieght"

File: common/app/redux/index.js (hunk 2 above)
Comment (Bouncey) [reply common/app/redux/index.js]: I keep getting a weird `not a function` error, when I logged `typeof createEventMetaCreator`, I got `undefined`

Changing it to a function declaration fixed that.

File: server/services/challenge.js (hunk 3 above)
Comment (Bouncey) [reply server/services/challenge.js]: Fixed

File: common/app/Map/Map.jsx (hunk 4 above)
Comment (Bouncey) [reply common/app/Map/Map.jsx]: Nice catch :+1:

**Comment by Bouncey** [comment]
@BerkeleyTrue I have created a `nextChallengeSelector` which is used in the new `fetchChallengesForNextBlockEpic` and the current `nextChallengeEpic`.

**Review by BerkeleyTrue — APPROVED** [review]

**Pull request merged**